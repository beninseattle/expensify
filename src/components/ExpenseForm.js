import React, {Component} from 'react';
import moment from 'moment';
// react dates must be initialized before importing
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import Expense from '../models/expense';

/**
 * @class ExpenseForm
 * @property {Object} props
 * @property {onSubmitExpenseCallback} props.onSubmit
 * @property {Expense} props.expense
 * @property {Object} state
 * @property {Expense} state.expense
 */
export default class ExpenseForm extends Component {
  onDescriptionChange = (e) => {
    let newExpense = new Expense(this.state.expense);
    newExpense.description = e.target.value;
    this.setState(() => ({expense: newExpense}));
  };
  onNoteChange = (e) => {
    let newExpense = new Expense(this.state.expense);
    newExpense.note = e.target.value;
    this.setState(() => ({expense: newExpense}));
  };
  // TODO: Fix this input
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      let newExpense = new Expense(this.state.expense);
      newExpense.amount = parseFloat(amount) * 100;
      this.setState(() => ({expense: newExpense}));
    }
  };

  /**
   * @param {moment} createdAt
   */
  onDateChange = (createdAt) => {
    if (createdAt) {
      let newExpense = new Expense(this.state.expense);
      newExpense.createdAt = createdAt.valueOf();
      this.setState(() => ({expense: newExpense}));
    }
  };
  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}));
  };
  handleSubmit = (e) => {
    e.preventDefault();

    if (!this.state.expense.isValid()) {
      this.setState(() => ({error: 'A description and amount is required'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit(this.state.expense);
    }
  };

  /**
   * @param {Object} props
   * @property {onSubmitExpenseCallback} props.onSubmit
   * @property {Expense} [props.expense]
   */
  constructor(props) {
    super(props);
    this.submitButton = 'Save changes';
    let expense = props.expense;
    if (expense === undefined) {
      expense = new Expense();
      this.submitButton = 'Save expense';
    }
    this.state = {
      expense: expense,
      calendarFocused: false,
      error: ''
    };
  };

  render() {
    return (
      <form className="form" onSubmit={this.handleSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          name="description"
          autoFocus
          value={this.state.expense.description}
          onChange={this.onDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          name="amount"
          value={this.state.expense.amountAsFloat()}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={moment(this.state.expense.createdAt)}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          name="note"
          placeholder="Add a note (optional)"
          value={this.state.expense.note}
          onChange={this.onNoteChange}/><br/>
        <div>
          <button className="button">{this.submitButton}</button>
        </div>
      </form>
    );

  }
}