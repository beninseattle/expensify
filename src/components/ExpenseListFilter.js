import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
// react-dates must be initialized first before importing
import 'react-dates/initialize';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, setStartDate, setStopDate, sortByAmount, sortByDate} from "../store/actions/filters";

// TODO: Set up type object for filters object?
/**
 * @property {Object} props
 * @property {function(string)} props.setTextFilter
 * @property {function(moment)} props.setStartDate
 * @property {function(moment)} props.setStopDate
 * @property {function} props.sortByDate
 * @property {function} props.sortByAmount
 * @property {Object} props.filters
 */
export class ExpenseListFilter extends React.Component {
  state = {
    calendarFocused: null
  };

  handleSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onDateChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setStopDate(endDate);
  };
  onFocusChange = calendarFocused => this.setState({calendarFocused});

  // TODO: Revisit usage of moment vs timestamps as it's unclear where which should be used and ugliness below
  render() {
    return (
      <div>
        Filter name
        <input type="text"
               value={this.props.filters.text}
               onChange={this.onTextChange}
        />
        <DateRangePicker
          startDate={this.props.filters.startDate ? moment(this.props.filters.startDate) : this.props.filters.startDate}
          endDate={this.props.filters.stopDate ? moment(this.props.filters.stopDate) : this.props.filters.stopDate}
          onDatesChange={this.onDateChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
        <select value={this.props.filters.sortBy} onChange={this.handleSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setStartDate: (date) => dispatch(setStartDate(date)),
  setStopDate: (date) => dispatch(setStopDate(date)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

const mapStateToProps = (state) => ({
  filters: state.filters
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);