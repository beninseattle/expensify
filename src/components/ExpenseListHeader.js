import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../store/selectors/expenses';
import selectExpensesTotal from '../store/selectors/expenses-total';

/**
 * @param {Object} props
 * @param {int} props.expensesCount
 * @param {int} props.expensesTotal
 * @constructor
 */
export const ExpenseListHeader = ({expensesCount, expensesTotal}) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
  const formattedTotal = (expensesTotal/100).toLocaleString(undefined, {style: 'currency', currency: 'USD'});
  return (
    <div className="list_header">
      <h2>{expensesCount} {expenseWord} found for a total of {formattedTotal}</h2>
    </div>
  )
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpenseListHeader);