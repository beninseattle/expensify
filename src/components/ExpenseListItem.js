import React from 'react';
import {Link} from 'react-router-dom';
import paths from '../routers/paths';

/**
 * @param {object} props
 * @param {number} props.index
 * @param {Expense} props.expense
 * @constructor
 */
const ExpenseListItem = ({index, expense}) => (
  <Link className="list-item" to={paths.editExpense + expense.id}>
    <div>
      <h3 className="list-item__title">{expense.description}</h3>
      <span className="list-item__subtitle">({expense.createdAtDateString()})</span>
    </div>
    <h3 className="list-item__data">{expense.amountCurrencyString()}</h3>
  </Link>
);

export default ExpenseListItem;