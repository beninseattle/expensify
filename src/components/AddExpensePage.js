import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../store/actions/expenses';
import paths from '../routers/paths';

/**
 * @property {Object} props
 * @property {mapDispatchToProps} props.onSubmit
 * @property {Object} props.history
 * @property {function(string)} props.history.push
 */
export class AddExpensePage extends React.Component {
  /**
   * @callback onSubmitExpenseCallback
   * @param expense
   */
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push(paths.dashboard);
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    );
  }
}

/**
 * @callback
 * @param {function(expenseAction)} dispatch
 * @returns {{addExpense: (function(Expense): *)}}
 */
const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense))
  };
};
export default connect(undefined, mapDispatchToProps)(AddExpensePage);