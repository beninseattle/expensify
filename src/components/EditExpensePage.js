import React from 'react';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import {editExpense, startDeleteExpense} from "../store/actions/expenses";
import paths from '../routers/paths';

/**
 * @property {Object} props
 * @property {Expense} props.expense
 * @property {function(Expense)} props.editExpense
 * @property {function(Expense)} props.deleteExpense
 * @property {Object} props.history
 * @property {function(string)} props.history.push
 * @property {function(string)} props.history.redirect
 */
export class EditExpensePage extends React.Component {
  /**
   * @callback onSubmitExpenseCallback
   * @param {Expense} expense
   */
  onSubmit = (expense) => {
    this.props.editExpense(expense);
    this.props.history.push(paths.dashboard);
  };
  /**
   * @callback onDeleteExpenseCallback
   */
  onDelete = () => {
    this.props.deleteExpense(this.props.expense);
    this.props.history.push(paths.dashboard);
  };

  render() {
    if( this.props.expense === undefined ) {
      return (
        <Redirect to={paths.dashboard} />
      );
    }

    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onDelete}>delete
        </button>
      </div>
    );
  }
}
/**
 *
 * @param {Object} state
 * @property {Expense[]} state.expenses
 * @param {Object} props
 * @property {string} props.match.params.id
 * @returns {{expense: T}}
 */
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.equals(props.match.params.id))
});
/**
 * @callback
 * @param {function(expenseAction)} dispatch
 * @returns {{addExpense: (function(Expense)), deleteExpense: (function(Expense))}}
 */

const mapDispatchToProps = (dispatch) => ({
  editExpense: (expense) => dispatch(editExpense(expense)),
  deleteExpense: (expense) => dispatch(startDeleteExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);