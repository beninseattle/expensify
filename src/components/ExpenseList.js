import React from 'react';
import {connect} from 'react-redux';
import ExpenseListHeader from './ExpenseListHeader';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../store/selectors/expenses';
import selectExpensesTotal from '../store/selectors/expenses-total';

/**
 * @param {Object} props
 * @param {Expense[]} props.expenses
 * @constructor
 */
export const ExpenseList = ({expenses}) => {
  if( expenses.length === 0 ) {
    return (
      <p>
        No expenses found.
      </p>
    );
  } else {
    return (
      <div>
        <ExpenseListHeader/>
        { expenses.map(
              (expense, index) =>
                <ExpenseListItem
                  key={expense.id}
                  index={index + 1}
                  expense={expense}
                />)
        }
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
