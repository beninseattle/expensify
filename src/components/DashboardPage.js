import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const DashboardPage = () => (
  <div>
    <ExpenseListFilter/>
    <ExpenseList/>
  </div>
);

export default DashboardPage;