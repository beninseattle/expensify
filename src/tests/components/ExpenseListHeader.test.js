import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseListHeader} from "../../components/ExpenseListHeader";

test('should render ExpenseListHeader for one expense', () => {
  const wrapper = shallow(<ExpenseListHeader expensesCount={1} expensesTotal={200}/>);

  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseListHeader for multiple expenses', () => {
  const wrapper = shallow(<ExpenseListHeader expensesCount={2} expensesTotal={450000}/>);

  expect(wrapper).toMatchSnapshot();
});
