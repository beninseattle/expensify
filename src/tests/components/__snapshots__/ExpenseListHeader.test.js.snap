// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListHeader for multiple expenses 1`] = `
<div
  className="list_header"
>
  <h2>
    2
     
    expenses
     found for a total of 
    US$ 4,500.00
  </h2>
</div>
`;

exports[`should render ExpenseListHeader for one expense 1`] = `
<div
  className="list_header"
>
  <h2>
    1
     
    expense
     found for a total of 
    US$ 2.00
  </h2>
</div>
`;
